Optional stage support

ajordens opened the pull request:
- Parameter or Expression based

ajordens commented on orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/OptionalStageSupport.groovy:
```diff
@@ -0,0 +1,91 @@
+
+    @Override
+    boolean isOptional(Stage stage) {
+      def augmentedContext = [:] + stage.context
```
Similar to what `ExpressionPreconditionTask` is doing.

ajordens commented:
Motivation behind this is to allow people to configure pipelines with particular stages being optional (_by parameter or expression_).

Optional stages should be skipped but any subsequent stages will be executed normally.

PTAL @spinnaker/netflix-reviewers / @spinnaker/google-reviewers

tomaslin commented on orca-core/src/main/groovy/com/netflix/spinnaker/orca/batch/ExecutionPropagationListener.groovy:
```diff
@@ -87,7 +87,7 @@ class ExecutionPropagationListener extends JobExecutionListenerSupport implement
     }
 
-    if (orcaTaskStatus == STOPPED) {
+    if (orcaTaskStatus?.isSuccessful()) {
```
suspect we would need to modify echo to bypass skipped stages as well. https://github.com/spinnaker/echo/blob/78500ab1bff77bf1dc1072e52bd00c2cada64043/echo-notifications/src/main/groovy/com/netflix/spinnaker/echo/notification/AbstractEventNotificationAgent.groovy#L70-L72

ajordens replied:
Good call ... don't really want to send notifications for `SKIPPED` stages.

tomaslin commented on orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/OptionalStageSupport.groovy:
```diff
@@ -0,0 +1,91 @@
+    boolean isOptional(Stage stage) {
+      def augmentedContext = [:] + stage.context
+      if (stage.execution instanceof Pipeline) {
+        augmentedContext.put('trigger', ((Pipeline) stage.execution).trigger)
```
I think we're doing this in 3 places, should move to a common utils class perhaps

tomaslin commented:
I think calling the field 'enabled' instead of 'optional' would be clearer, otherwise LGTM

ajordens commented:
@tomaslin Thoughts some some sort of a prefix to denote context keys as being Spinnaker-only? I worry about something like `optional` or `enabled` in that they're pretty fairly and could just as easily be a result of a `jenkins` stages which would (albeit edge-casey) cause confusion.

tomaslin commented:
stageEnabled

ajordens commented:
Was thinking the same thing ... aligns with `stageTimeoutMs`.

cfieber commented:
think this is missing a change in `WaitForRequisiteStage` to handle the skipped state (@zanthrash just added the `FAILED_CONTINUE` state in there in #830)

cfieber commented on orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/OptionalStageSupport.groovy:
```diff
@@ -0,0 +1,91 @@
+  /**
+   * An {@link OptionalStageEvaluator} that will perform a case-insensitive comparison against a pipeline parameter.
+   */
+  private static class ParameterOptionalStageEvaluator implements OptionalStageEvaluator {
```
thoughts on special casing this instead of just using the equivalent expression? `parameters.key == value` as the expression doesn't seem much worse and saves having an extra case both here and in deck

ajordens replied:
That's fair ... chalk it up to starting with the parameter based implementation and tests.

About the only thing it gives you is perhaps a little more flexibility around being case-insensitive on matching. 

I'll strip it out for now.

ajordens commented:
- removed the explicit parameter support (for now)
- added `SKIPPED` to `WaitForRequisiteStage`
- moved the `augmentedContext` creation to `ContextParameterProcessor#buildExecutionContext` and retrofitted existing usages
- suppressed events for SKIPPED and CANCELED statuses
